README
README for xsysinfo version 1.6.


WHAT IS XSYSINFO:

  Xsysinfo is an X application to display some Linux kernel parameters in
  graphical form. It is like a mix of top, free and xload with the difference
  that the values are shown in form of a horizontal bar. The displayed values
  are: CPU load average, CPU load, memory and swap sizes (details see below).

  Xsysinfo 1.6 runs at least under Linux 0.99pl5 with the proc filesystem (as
  kernels from this release on support the CPU idle time, no patch is
  necessary like in earlier versions of xsysinfo). Starting with version 1.3,
  xsysinfo also supports the new /proc/meminfo-format used in kernels >= 1.3.x


FOR THE IMPATIENT:

  If you want to take a look at xsysinfo, before you decide to install it,
  make sure the proc fs is mounted at /proc, then just type:

	XAPPLRESDIR=. xsysinfo &


HOW TO INSTALL:

  There is a compiled ELF-version of xsysinfo included in the archive, which
  was built under RedHat 4.2 with XFree86 3.3.1 and libc 5.3.12. If you want
  to install this binary just type

	make install

  as root. xsysinfo runs under normal (755) permissions, so you don't have to
  make it setuid/setgid.

  NOTE: Use this way to install xsysinfo! `make install' not only copies the 
  binary to the right place, it also installs the system wide application 
  defaults file (XSysinfo). Without this file you'll get a running xsysinfo,
  but it won't look very nice... 

  If you want to create your own compiled version just type:

	xmkmf
	make clean
	make

  (you'll need the programming package and the libraries from the XFree
  distribution to do this.)
  Now you can install it as described above.

  Xsysinfo requires that the proc filesystem is compiled into the kernel
  (this is the default anyway) and that it is mounted on directory /proc.
  To check this just do an `ls' on /proc, which should give something like
  this:

	$ ls /proc
	1/       21/      27/      39/      42/      52/      kmsg     uptime
	14/      22/      28/      4/       44/      75/      loadavg
	15/      23/      29/      40/      46/      87/      meminfo
	20/      24/      38/      41/      47/      self/    version

  To mount the proc filesystem, add the following line to your /etc/fstab:

  none            /proc           proc    defaults        # proc fs


WHAT IS DISPLAYED:

  Xsysinfo displays the following values:

    - CPU load average
	CPU load average between 0.000-8.000. The gauge's bar is subdivided
	into segments, where one segment represents a load value of 1.0. The
	bar's full length is automatically scaled, depending on the displayed
	value:

		load value (l)		full length	# segments
		------------------------------------------------------
		0.00 <= l < 1.00	1.00		1
		1.00 <= l < 2.00	2.00		2
		2.00 <= l < 3.00	3.00		3
		... and so on ...

  - CPU load
	percentage CPU load time to CPU idle time subdivided in three segments:
	user load, system load and nice load.

  - Memory
	The memory gauge's bar is subdivided into two segments with following
	values (from left to right):

	o  amount of physical memory, which is used by processes

	o  physical memory used for the page and buffer cache

	The length of the whole bar, which is the sum of these two values,
	shows the amount of physical memory currently used by the system.

  - Swap
	The percentage of swap space used by the system to total amount of
	swap space.


COMMAND LINE OPTIONS:

  Xsysinfo accepts the following command line options:

	-update <n>                     set update rate to <n> milli-seconds
	-title   (-notitle)             show (don't show) title string
	-labels  (-nolabels)            show (don't show) gauge labels
	-loadavg (-noloadavg)           show (don't show) CPU load average value
	-load    (-noload)              show (don't show) CPU load value
	-mem     (-nomem)               show (don't show) Memory info
	-swap    (-noswap)              show (don't show) Swap info
	-help                           show this message

  You can also specify any of the standard toolkit options (see: manpage X(1)).


CONFIGURATION:

  As the most X programs, xsysinfo is highly configurable. There are two
  resource files: XSysinfo with color independent settings and XSysinfo-color
  with color resources. The latter is automatically loaded, if your server 
  is set up to use color customized resource files. To do this just add 

  *customization: -color

  in your server resource file loaded by xrdb at startup (usually 
  /usr/X11R6/lib/X11/xinit/.Xresources).


ACKNOWLEDGMENTS:

  Thanks to Linus and the other kernel hackers, who made Linux to this great
  OS. Also thanks to Muhammad M. Saggaf for his suggestions and ideas for this
  xsysinfo release. And last but not least thanks to Lucas Bussmann and 
  Joerg Fries for creating the color resources.


COMMENTS:

  Xsyinfo is written by Gabor Herr (herr@iti.informatik.th-darmstadt.de) and
  currently maintained by Ronald Wahl (ronald.wahl@informatik.tu-chemnitz.de).
  Comments, bug reports, whishes etc... are welcome.

  Have fun watching the kernel...

Gabor & Ronald

5/4/99 Mike Mondragon <dragon@gocougs.wsu.edu> added SMP support to version 1.7


CHANGES
Changes from version 1.6 to 1.7:
o  Added full SMP support as follows:

o  the follow data and/or functions were modified to support a SMP
   kernel in sysinfo.h
   char *cpu field added to load struct		// modified
   extern int get_total_procs( void );		// added

o  the follow data and/or functions were modified to support a SMP
   kernel in sysinfo.c
   void get_load(struct load * result)		// modified
   int get_total_procs( void );			// added

o  the follow data and/or functions were modified to support a SMP 
   kernel in xsysinfo.c
   struct AppData, *AppDataPtr			// modified
   static XtResource application_resources[]	// modified
   static XrmOptionDescRec options[]		// modified
   static void getloadCB( widget, client_data, call_data ) // modified
   static void create_app( top )		// modified

o  Added a define to allow turning off of the now default SMP mode.
   Turning off SMP mode on a SMP enabled kernel will display the
   the average of all the cpu's load which is still contained on
   on the first line of /proc/stat

   Mike Mondragon <dragon@gocougs.wsu.edu>

Changes from version 1.5 to 1.6:

o  Added a define to allow specifying the update time in milli-seconds. This
   is currently the default. To change the behaviour edit the Imakefile and
   XSysinfo.ad, call `imake', `make clean' and `make'.
   This change was contributed by Thomas M. Ott <Labalutsch@aol.com>. 

o  Added support for /proc/swaps in newer (2.1.24) kernels. This pseudo file
   gives information about every swapfile used by the system.
   This change was contributed by Florian Depke <depke@cs.tu-berlin.de>.

o  Fixed a cosmetical bug which leaves space between the right end of the bar
   and the right end of the gauge if the bar should have the same width as the
   gauge.

o  There was a bug which cause a segmentation fault after starting xsysinfo
   if compiled without optimization. This is now fixed.

o  minor changes to the X resources

o  The changes in xsysinfo are now explained in a separate CHANGES file.

o  Added "Maintained-by:" to the LSM file.

   Ronald Wahl <Ronald.Wahl@informatik.tu-chemnitz.de>

Changes from version 1.4 to 1.5:

o  The mem-bar now consist of two segments only: used memory and cache/buffers.
o  Rename the load to loadavg.
o  Replaced the idle bar with a load bar consisting of three segments:
   user load/system load/nice load
o  Added a define to allow empty segments to disappear (currently the default).
   Otherwise every segment will be at least 1 pixel wide. To change the
   behaviour edit the Imakefile, call `imake', `make clean' and `make'.
o  Changed color of some bars.

   Ronald Wahl <Ronald.Wahl@informatik.tu-chemnitz.de>

Changes from version 1.3 to 1.4:

o  Fixed mem-bar that was broken in 1.3. Removed 4th segment again and
   merged buffers & cached.
o  Added -nolabels switch for use with fvwm's GoodStuff (thanks to Ronald
   Wahl <rwa@informatik.tu-chemnitz.de> for patches)

   Martin Buck   <martin-2.buck@student.uni-ulm.de>


Changes from version 1.2 to 1.3:

o  Made xsysinfo understand new /proc/meminfo format introduced in 1.3.x
   kernels.

o  Split mem-bar into 4 segments (data, shared, buffers, cached) if info
   is available.

   Martin Buck   <martin-2.buck@student.uni-ulm.de>


Changes from version 1.1 to 1.2:

o	load value

   The gauge for CPU load is now capable to display values up to 8.0. The 
   bar is automatically subdivided in segments when necessary (see. below). 


o	proc filesystem

   The proc filesystem under kernel 0.99pl5 reports the absolute CPU idle
   time in file /proc/uptime and not like the earlier kernels, with my patches
   applied, in file /proc/idle. I changed xsysinfo to obtain the idle value 
   from uptime.


o	swap display

   The 0.99pl5 kernel doesn't support a detailed swap space reporting in
   /proc/meminfo any longer. Only the total swap space statistics can be 
   obtained, so the swap gauge of xsysinfo is not segmented as in version 1.1.


o	color resources

   There is a separate resource file for color displays, which is loaded
   automatically by X's customizations mechanism (see below).


Changes from version 1.0 to 1.1:

o	proc filesystem

   I have completely rewritten the code to get the kernel params. It now uses
   the proc filesystem, which made the whole stuff a lot easier. Another
   advantage of the proc fs is, that xsysinfo can run under normal permissions
   and need not to be recompiled with every new kernel release. 


o	memory display changed

   The two gauges for memory and cache sizes are now merged together. The
   new gauge is still in form of a bar, but is subdivided into three `segments',
   each displayed in a different color. The meaning of the three segments
   is: buffer cache, shared part of memory, non-shared part of memory. The 
   length of the whole bar is proportional to the amount of all used memory.


o	more than one swap area supported

   The gauge for swap size is now also subdivided, one `segment' for each
   swap file/partition.


o	gauges now can be disabled

   Now it is possible to disable each of the gauges either through 
   application resources or through command line options.


sysinfo.c

/*
 * sysinfo.c		- get kernel info
 *
 */

/* 
 * Written by Gabor Herr <herr@iti.informatik.th-darmstadt.de>.
 *
 * Copyright (c) 1992, 1993 by Gabor Herr, all rights reserved.
 * 
 * Permission to use, copy, modify, distribute, and sell this software
 * and its documentation for any purpose is hereby granted without fee,
 * provided that the above copyright notice appear in all copies and
 * that both that copyright notice and this permission notice appear in
 * supporting documentation, and that may name is not used in
 * advertising or publicity pertaining to distribution of the software
 * without specific, written prior permission. I make no representations
 * about the suitability of this software for any purpose. It is
 * provided "as is" without express or implied warranty.
 */

/* $Id: sysinfo.c,v 1.3 1993/02/12 11:40:16 gabor Exp $ */
 
#include "sysinfo.h"
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>

static int meminfo_fd;
static int load_fd;
static int loadavg_fd;
static int swaps_fd;

static char buffer[1024];

static void reread( int fd, char *message )
{
	if ( lseek( fd, 0L, 0 ) == 0 &&
	     read( fd, buffer, sizeof(buffer) - 1 ) > 0 )
		return;
       
	perror(message);
	exit(1);
}

static int getentry(const char *tag, const char *bufptr) {
	char *tail;
	int retval, len = strlen(tag);
  
	while (bufptr) {
		if (*bufptr == '\n') bufptr++;
		if (!strncmp(tag, bufptr, len)) {
			retval = strtol(bufptr + len, &tail, 10);
			if (tail == bufptr + len) {
				return -1;
			} else {
				return retval;
			}
		}
		bufptr = strchr( bufptr, '\n' );
	}
	return -1;
}

void get_meminfo( int *swapdevs, struct meminfo *result )
{
	int i;
	char *bufptr;
	int res;
	int sw_size, sw_used;

	reread( meminfo_fd, "get_meminfo");

	/* Try new /proc/meminfo format first */
	*swapdevs = 1;
	if ((result[0].total   = getentry("MemTotal:", buffer))  < 0 ||
	    (result[0].free    = getentry("MemFree:", buffer))   < 0 ||
	    (result[0].shared  = getentry("MemShared:", buffer)) < 0 ||
	    (result[0].buffers = getentry("Buffers:", buffer))   < 0 ||
	    (result[0].cache   = getentry("Cached:", buffer))    < 0 ||
	    (result[1].total   = getentry("SwapTotal:", buffer)) < 0 ||
	    (result[1].free    = getentry("SwapFree:", buffer))  < 0) {

		/* Fall back to old format */
		bufptr = strchr( buffer, '\n' ) + 1;
		sscanf( bufptr, "Mem: %d %*d %d %d %d",
			&result[0].total,
			&result[0].free,
			&result[0].shared,
			&result[0].cache );
		result[0].buffers = -1;
    
		for ( i = 1; i < MAX_SWAPFILES; i++ ) {

			bufptr = strchr( bufptr, '\n' ) + 1;

			if ( *bufptr == '\0' ||
			     (res = sscanf( bufptr, "Swap: %d %*d %d",
					    &result[i].total,
					    &result[i].free )) != 2 )
				break; 

		}

		*swapdevs = i - 1;  
	} else if (swaps_fd > 0) {

		reread( swaps_fd, "get_meminfo");
		bufptr = buffer;

		for ( i = 1; i < MAX_SWAPFILES; i++ ) {

			bufptr = strchr( bufptr, '\n' ) + 1;
			if ( *bufptr == '\0' ||
			     (res = sscanf( bufptr, "%*s %*s %d %d",
					    &sw_size, &sw_used)) != 2 )
				break;

			result[i].total = sw_size;
			result[i].free = sw_size - sw_used;
		}

		*swapdevs = i - 1;  
	}
}

double get_loadavg( void )
{
	double load;

	reread( loadavg_fd, "get_loadavg");

	sscanf( buffer, "%lf", &load );

	return load;
}

void get_load(struct load * result)
{
	static int initialized = 0;
	static struct load *last_load;
	struct load *loadptr;
	struct load curr_load;
	int procs, i, index;
	char *bufptr;

	if(!initialized) {
		procs = get_total_procs();
		last_load = (struct load *)malloc(procs * sizeof(struct load));
		loadptr = last_load;
		for( i = 0; i < procs; i++) {
			loadptr->total = 0;
			loadptr->user = 0;
			loadptr->system = 0;
			loadptr->nice = 0;
			loadptr->idle = 0;
			loadptr->cpu = NULL;
			loadptr++;
		}
		initialized = 1;
	}

	reread( load_fd, "get_load" );
	bufptr = buffer;

	if(sscanf(result->cpu, "cpu%d", &index) != 1)
		loadptr = last_load;
	else
		loadptr = last_load + index;

	bufptr = strstr(bufptr, result->cpu);
	sscanf( bufptr, "%*s %lu %lu %lu %lu\n", &curr_load.user,
		&curr_load.nice, &curr_load.system, &curr_load.idle );
	curr_load.total = curr_load.user + curr_load.nice + curr_load.system +
		curr_load.idle;

	result->total  = curr_load.total  - loadptr->total;
	result->user   = curr_load.user   - loadptr->user;
	result->nice   = curr_load.nice   - loadptr->nice;
	result->system = curr_load.system - loadptr->system;
	result->idle   = curr_load.idle   - loadptr->idle;
	loadptr->total  = curr_load.total;
	loadptr->user   = curr_load.user;
	loadptr->nice   = curr_load.nice;
	loadptr->system = curr_load.system;
	loadptr->idle   = curr_load.idle;
}

int sysinfo_init( void )
{
	if ((meminfo_fd = open("/proc/meminfo",O_RDONLY)) < 0) {
		perror("/proc/meminfo");
		return 1;
	}
	if ((loadavg_fd = open("/proc/loadavg",O_RDONLY)) < 0) {
		perror("/proc/loadavg");
		return 1;
	}
	if ((load_fd = open("/proc/stat",O_RDONLY)) < 0) {
		perror("/proc/stat");
		return 1;
	}
	/* Failing here is not critical. We can get the info from meminfo. */
	swaps_fd = open("/proc/swaps",O_RDONLY);

	return 0;
}

int get_total_procs( void )
{
	int count = 1;
	char cpu_num[7];
	char *bufptr;

	reread( load_fd, "get_total_procs" );

	while(1)
	{
		bufptr = buffer;
		sprintf(cpu_num,"cpu%d ", count);
		if(strstr(bufptr, cpu_num) != NULL)
			count++;
		else
			return count;
	}

	return -1;
}


sysinfo.h
/*
 * sysinfo.h		- get kernel info
 *
 */

/* 
 * Written by Gabor Herr <herr@iti.informatik.th-darmstadt.de>.
 *
 * Copyright (c) 1992, 1993 by Gabor Herr, all rights reserved.
 * 
 * Permission to use, copy, modify, distribute, and sell this software
 * and its documentation for any purpose is hereby granted without fee,
 * provided that the above copyright notice appear in all copies and
 * that both that copyright notice and this permission notice appear in
 * supporting documentation, and that may name is not used in
 * advertising or publicity pertaining to distribution of the software
 * without specific, written prior permission. I make no representations
 * about the suitability of this software for any purpose. It is
 * provided "as is" without express or implied warranty.
 */

/* $Id: sysinfo.h,v 1.4 1993/02/12 11:40:16 gabor Exp $ */
 
#ifndef SYSINFO_INCLUDED
#define SYSINFO_INCLUDED

#define MAX_SWAPFILES 8

struct load {
	unsigned long total;
	unsigned long user;
	unsigned long system;
	unsigned long nice;
	unsigned long idle;
	char *cpu;
};

struct meminfo {
	int total;
	int cache;
	int buffers;
	int free;
	int shared;
};

extern void get_meminfo( int *, struct meminfo * );
extern double get_loadavg(void);
extern void get_load(struct load *);
extern int sysinfo_init(void);
extern int get_total_procs( void );

#endif


BarGauge.h

#ifndef _XgBarGauge_h
#define _XgBarGauge_h

/***********************************************************************
 *
 * BarGauge: Widget to periodically display a value in form of a horizontal bar.
 * based on Athena StripChart Widget.
 *
 * Author: Gabor Herr		herr@iti.informatik.th-darmstadt.de
 *
 * $Id: BarGauge.h,v 1.3 1993/02/07 23:54:15 gabor Exp $
 *
 ***********************************************************************/

/***********************************************************
Copyright 1992, 1993 by Gabor Herr
Copyright 1987, 1988 by Digital Equipment Corporation, Maynard, Massachusetts,
and the Massachusetts Institute of Technology, Cambridge, Massachusetts.

                        All Rights Reserved

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose and without fee is hereby granted, 
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in 
supporting documentation, and that the names of Gabor Herr or Digital or MIT not
be used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.  

GABOR HERR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
GABOR HERR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

******************************************************************/

/* 

 Name		     Class		RepType		Default Value
 ----		     -----		-------		-------------

 BarGauge resources:

 autoAdjust	     AutoAdjust		Boolean		False
   If False, displays each value returned by getValues procedure in a
   separate segment. If True, only the first value returned by getValues
   is displayed in n segments, where n is value/maxValue; the first n-1
   segments are of length maxValue; the n'th segment is of
   length value%maxValue; the bar's total length is n*maxValue.
 
 segments	     Segments		int		1
   Total number of segments the bar is subdivided.
 
 segmentGap	     SegmentGap		int		2
   Gap between adjacent segments in pixel.
 
 foreground	     Foreground		Pixel		XtDefaultForeground
 foreground1	     Foreground		Pixel		XtDefaultForeground
 foreground2	     Foreground		Pixel		XtDefaultForeground
 foreground3	     Foreground		Pixel		XtDefaultForeground
 foreground4	     Foreground		Pixel		XtDefaultForeground
 foreground5	     Foreground		Pixel		XtDefaultForeground
 foreground6	     Foreground		Pixel		XtDefaultForeground
 foreground7	     Foreground		Pixel		XtDefaultForeground
   Foreground color of each of the max. 8 segments.
 
 getValue	     Callback		XtCallbackList	NULL
   Callback procedure to be called to obtain the actual segment values. The
   call_data parameter is a pointer to XgBarGaugeValues.
 
 maxValue	     MaxValue		float		1.0
   The bar's full length. I.e. a segment with this value will completely fill
   the gauge.
 
 update		     Interval		int		10 (seconds)
   Time between two calls of the getValue procedure.

 
 Inherited resources:
 
 accelerators	     Accelerators	AcceleratorTable NULL
 ancestorSensitive   AncestorSensitive	Boolean		True 
 background	     Background		Pixel		XtDefaultBackground
 backgroundPixmap    Pixmap		Pixmap		XtUnspecifiedPixmap
 borderColor	     BorderColor	Pixel		XtDefaultForeground
 borderPixmap	     Pixmap		Pixmap		XtUnspecifiedPixmap
 borderWidth	     BorderWidth	Dimension	1
 colormap	     Colormap		Colormap	parent's colormap
 cursor		     Cursor		Cursor		None
 cursorName	     Cursor		String		NULL
 depth		     Depth		int		parent's depth
 destroyCallback     Callback		XtCallbackList	NULL
 height		     Height		Dimension	120
 insensitiveBorder   Insensitive	Pixmap		GreyPixmap
 mappedWhenManaged   MappedWhenManaged	Boolean		True
 pointerColor	     Foreground		Pixel		XtDefaultForeground
 pointerColorBackground Background	Pixel		XtDefaultBackground
 screen		     Screen		Screen		parent's screen
 sensitive	     Sensitive		Boolean		True
 translations	     Translations	TranslationTable NULL
 width		     Width		Dimension	120
 x		     Position		Position	0
 y		     Position		Position	0

*/

#ifndef _XtStringDefs_h_
#define XtNupdate "update"
#endif

#define XtNautoAdjust "autoAdjust"
#define XtCAutoAdjust "AutoAdjust"

#define XtNgetValue "getValue"

#define XtNsegments "segments"
#define XtCSegments "Segments"

#define XtNsegmentGap "segmentGap"
#define XtCSegmentGap "SegmentGap"

#define XtNmaxValue "maxValue"
#define XtCMaxValue "MaxValue"

#define XtNforeground1 "foreground1"
#define XtNforeground2 "foreground2"
#define XtNforeground3 "foreground3"
#define XtNforeground4 "foreground4"
#define XtNforeground5 "foreground5"
#define XtNforeground6 "foreground6"
#define XtNforeground7 "foreground7"

/* call data param for getValue */
typedef struct {
	Cardinal num_values;		/* # of values */
	double *values;		/* values[i] = value for i-th bar segment */
} XgBarGaugeValues;

typedef struct _BarGaugeRec *BarGaugeWidget;
typedef struct _BarGaugeClassRec *BarGaugeWidgetClass;

extern WidgetClass barGaugeWidgetClass;

#endif /* _XgBarGauge_h */


xsysinfo.c
/*
 * xsysinfo: X application for displaying some kernel parameters
 *
 * $Id: xsysinfo.c,v 1.9 1993/02/12 11:40:16 gabor Exp $
 *
 */

/* 
 * Written by Gabor Herr <herr@iti.informatik.th-darmstadt.de>.
 *
 * Copyright (c) 1992, 1993 by Gabor Herr, all rights reserved.
 * 
 * Permission to use, copy, modify, distribute, and sell this software
 * and its documentation for any purpose is hereby granted without fee,
 * provided that the above copyright notice appear in all copies and
 * that both that copyright notice and this permission notice appear in
 * supporting documentation, and that may name is not used in
 * advertising or publicity pertaining to distribution of the software
 * without specific, written prior permission. I make no representations
 * about the suitability of this software for any purpose. It is
 * provided "as is" without express or implied warranty.
 */

#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Label.h>
#include "BarGauge.h"
#include <stdio.h>
#include "sysinfo.h"

#define norm(x,max)	((max) == 0 ? 0 : (double)(x) / (max))

/* Application Ressources: */
typedef struct {
	Boolean show_title;
	Boolean show_labels;
	Boolean show_loadavg;
	Boolean show_load;
	Boolean show_mem;
	Boolean show_swap;
	Boolean smp_kernel;
} AppData, *AppDataPtr;

static AppData appres;

static XtResource application_resources[] = {
	{ "showTitle", "ShowTitle", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_title), XtRImmediate, (XtPointer) True },
	{ "showLabels", "ShowLabels", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_labels), XtRImmediate, (XtPointer) True },
	{ "showLoadAvg", "ShowLoadAvg", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_loadavg), XtRImmediate, (XtPointer) True },
	{ "showLoad", "ShowLoad", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_load), XtRImmediate, (XtPointer) True },
	{ "showMem", "ShowMem", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_mem), XtRImmediate, (XtPointer) True },
	{ "showSwap", "ShowSwap", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,show_swap), XtRImmediate, (XtPointer) True },
	{ "smpKernel", "SmpKernel", XtRBoolean, sizeof(Boolean),
	  XtOffset(AppDataPtr,smp_kernel), XtRImmediate, (XtPointer) True }
};

/* Command line options */
static XrmOptionDescRec options[] = {

	{ "-update", "*.bar.update", XrmoptionSepArg, NULL },

	{ "-title",   "*showTitle", XrmoptionNoArg, (XtPointer) "True" },
	{ "-notitle", "*showTitle", XrmoptionNoArg, (XtPointer) "False" },

	{ "-labels",   "*showLabels", XrmoptionNoArg, (XtPointer) "True" },
	{ "-nolabels", "*showLabels", XrmoptionNoArg, (XtPointer) "False" },

	{ "-loadavg",   "*showLoadAvg", XrmoptionNoArg, (XtPointer) "True" },
	{ "-noloadavg", "*showLoadAvg", XrmoptionNoArg, (XtPointer) "False" },

	{ "-load",   "*showLoad", XrmoptionNoArg, "True" },
	{ "-noload", "*showLoad", XrmoptionNoArg, "False" },

	{ "-mem",   "*showMem", XrmoptionNoArg, (XtPointer) "True" },
	{ "-nomem", "*showMem", XrmoptionNoArg, (XtPointer) "False" },

	{ "-swap",   "*showSwap", XrmoptionNoArg, (XtPointer) "True" },
	{ "-noswap", "*showSwap", XrmoptionNoArg, (XtPointer) "False" },

	{ "-smp",   "*smpKernel", XrmoptionNoArg, (XtPointer) "True" },
	{ "-nosmp", "*smpKernel", XrmoptionNoArg, (XtPointer) "False" }
};

static String fallback_resources[] = {
	"*.title.label:		Linux System Info",
	"*.loadavg*label:	CPU Load Avg:",
	"*.load*label:		CPU Load:",
	"*.mem*label:		Memory:  ",
	"*.swap*label:		Swap:    ",
	NULL
};

static XtAppContext context;
static struct meminfo swap[MAX_SWAPFILES];
static int swapfiles;
static int memupdate_freq = 0;

static void getloadavgCB( widget, client_data, call_data )
	Widget widget;
XtPointer client_data;
XtPointer call_data;
{
	XgBarGaugeValues *gauge = (XgBarGaugeValues *) call_data;
	gauge->values[0] = get_loadavg();  
}

static void getloadCB( widget, client_data, call_data )
	Widget widget;
XtPointer client_data;
XtPointer call_data;
{
	struct load load;
	XgBarGaugeValues *gauge = (XgBarGaugeValues *) call_data;

	load.cpu = (char *)malloc(7*sizeof(char));
	strcpy(load.cpu, (char *)client_data);
	get_load(&load);

	gauge->values[0] = norm(load.user, load.total);
	gauge->values[1] = norm(load.system, load.total);
	gauge->values[2] = norm(load.nice, load.total);
}

static void update_meminfo()
{
	static int count = 0;

	if ( count-- > 0 )
		return;

	count = memupdate_freq;
  
	get_meminfo( &swapfiles, swap );
}

static void getmemCB( widget, client_data, call_data )
	Widget widget;
XtPointer client_data;
XtPointer call_data;
{
	XgBarGaugeValues *gauge = (XgBarGaugeValues *) call_data;
	struct meminfo *mem = swap;

	update_meminfo();
  
	if (mem->buffers >= 0) {
		/* New style /proc/meminfo format */
    
		gauge->values[0] = norm(mem->total - mem->free - mem->cache -
					mem->buffers,mem->total);
		gauge->values[1] = norm(mem->cache + mem->buffers,mem->total);
	} else {
		/* Old style /proc/meminfo format (no buffers-entry) */
		gauge->values[0] = norm(mem->total - mem->free - mem->cache,
					mem->total);
		gauge->values[1] = norm(mem->cache,mem->total);
	}
}

static void getswapCB( widget, client_data, call_data )
	Widget widget;
XtPointer client_data;
XtPointer call_data;
{
	XgBarGaugeValues *gauge = (XgBarGaugeValues *) call_data;
	int i;
	int total;

	update_meminfo();
  
	if ( gauge->num_values != swapfiles )
		XtVaSetValues( widget, XtNsegments, swapfiles, NULL);

	for ( total = i = 0; i < gauge->num_values; i++ ) 
		total += swap[i + 1].total;

	for ( i = 0; i < gauge->num_values; i++ ) 
		gauge->values[i] = norm(swap[i + 1].total - swap[i + 1].free,
					total);
}

static Widget create_gauge( name, parent, segs, valueproc, client_data )
	String name;
Widget parent;
XtCallbackProc valueproc;
int segs;
XtPointer client_data;
{
	Widget gauge;
	Widget bar;
	Widget label;
  
	gauge = XtCreateManagedWidget( name, formWidgetClass, parent, NULL, 0 );
 
	if (appres.show_labels) {
		label = XtVaCreateManagedWidget( "name",labelWidgetClass,gauge,
						 XtNright, XawChainLeft,
						 NULL);

		bar = XtVaCreateManagedWidget( "bar",barGaugeWidgetClass,gauge,
					       XtNsegments, segs,
					       XtNfromHoriz, label,
					       XtNleft, XawChainLeft,
					       NULL);
	} else {
		bar = XtVaCreateManagedWidget( "bar",barGaugeWidgetClass,gauge,
					       XtNsegments, segs,
					       XtNleft, XawChainLeft,
					       NULL);
	}

	XtAddCallback( bar, XtNgetValue, valueproc, client_data );

	return bar;
}

static void create_app( top )
	Widget top;
{
	Widget form;
	Widget above_widget = NULL;
	Widget this;
	Widget bar;
	char *bufptr;
	int total_procs, i;
  
	form = XtCreateManagedWidget( "form", formWidgetClass, top, NULL, 0 );

	if (appres.show_title)
		above_widget = XtVaCreateManagedWidget( "title",
							labelWidgetClass, form,
							XtNright, XawChainRight,
							NULL);

	if (appres.show_loadavg)	{
		bar  = create_gauge( "loadavg", form, 1, getloadavgCB, NULL );
		XtVaSetValues( bar,
			       XtNautoAdjust, True,
			       NULL );
		this = XtParent(bar);
		XtVaSetValues( this,
			       XtNfromVert, above_widget,
			       XtNright,    XawChainRight,
			       NULL );
		above_widget = this;
	}
    
	if (appres.show_load) {
		total_procs = get_total_procs();
		if(!appres.smp_kernel || total_procs == 1) {
			bufptr = (char *)malloc(7*sizeof(char));
			sprintf(bufptr, "cpu ");
			this = XtParent(create_gauge( "load", form, 3, getloadCB, bufptr ));
			XtVaSetValues( this,
				       XtNfromVert, above_widget,
				       XtNright,    XawChainRight,
				       NULL );
			above_widget = this;
		}
		else {
			for( i = 0; i < total_procs; i ++) {
				bufptr = (char *)malloc(7*sizeof(char));
				sprintf(bufptr, "cpu%d ", i);

				this = XtParent(create_gauge( "load", form, 3, getloadCB, bufptr ));
				XtVaSetValues( this,
					       XtNfromVert, above_widget,
					       XtNright,    XawChainRight,
					       NULL );
				above_widget = this;
			}
		}
	}
    
	if (appres.show_mem)	{		
		this = XtParent(create_gauge( "mem", form, 2, getmemCB, NULL ));
		XtVaSetValues( this,
			       XtNfromVert, above_widget,
			       XtNright,    XawChainRight,
			       NULL );
		above_widget = this;
		memupdate_freq++;
	}
    
	if (appres.show_swap && swapfiles > 0)	{		
		this = XtParent(create_gauge( "swap", form, swapfiles,
					      getswapCB, NULL ));
		XtVaSetValues( this,
			       XtNfromVert, above_widget,
			       XtNright,    XawChainRight,
			       NULL );
		above_widget = this;
		memupdate_freq++;
	}

	if (above_widget == NULL)
		XtAppError( context,"All gauges disabled, nothing to show.\n" );
}

static void show_usage()
{
	fprintf(stderr,"xsysinfo version 1.7\n\n");
	fprintf(stderr,"usage: xsysinfo [options...]\n\n");
	fprintf(stderr,"where options include:\n");
#ifdef USE_MILLI_SECONDS       /* (ThMO) */
	fprintf(stderr,"-update <n>                     set update rate to <n> milli-seconds\n");
#else
	fprintf(stderr,"-update <n>                     set update rate to <n> seconds\n");
#endif
	fprintf(stderr,"-title   (-notitle)		show (don't show) title string\n");
	fprintf(stderr,"-labels  (-nolabels)		show (don't show) gauge labels\n");
	fprintf(stderr,"-loadavg (-noloadavg)		show (don't show) CPU load average value\n");
	fprintf(stderr,"-load    (-noload)		show (don't show) CPU load value\n");
	fprintf(stderr,"-mem     (-nomem)		show (don't show) Memory info\n");
	fprintf(stderr,"-swap    (-noswap)		show (don't show) Swap info\n");
	fprintf(stderr,"-smp     (-nosmp)               smp (not smp) kernel\n");
	fprintf(stderr,"-help				show this message\n");
	fprintf(stderr,"\nor one of the standard toolkit options.\n\n");
}

main( argc, argv )
	int argc;
char **argv;
{
	Arg args[1];
	Widget toplevel;

	toplevel = XtAppInitialize( &context, "XSysinfo",
				    options, XtNumber(options),
				    &argc, argv,
				    fallback_resources,
				    NULL, 0 );

	if (argc != 1) {
		show_usage();
		exit(1);
	}
			      
	XtGetApplicationResources( toplevel, &appres,
				   application_resources,
				   XtNumber( application_resources ),
				   NULL, 0 );
			      
	if (sysinfo_init() != 0) {
		fprintf(stderr,	"Proc filesystem isn't mounted on `/proc'.\n");
		exit(1);
	}
	update_meminfo();

	create_app( toplevel );

	XtRealizeWidget(toplevel);
	XtAppMainLoop(context);
}
